File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');
const N = Number(input[0]);
const info = input.slice(1).map((x) => x.split(' ').map(Number));
const graph = Array.from({length: N + 1}, () => []);
const parents = Array(N + 1).fill(0);

// 연결된 노드를 그래프로 만들기
for(let i = 0; i < N - 1; i++) {
    const [a, b] = info[i];
    graph[a].push(b);
    graph[b].push(a);
}

const queue = [1];
parents[1] = -1; // 루트는 부모 없음 표시

while (queue.length) {
  const node = queue.shift();

  for (const next of graph[node]) {
    if (parents[next] === 0) { // 방문 안 했으면
      parents[next] = node;    // 부모 기록
      queue.push(next);
    }
  }
}

// 2번부터 출력
for (let i = 2; i <= N; i++) {
  console.log(parents[i]);
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
from collections import deque

input = sys.stdin.read().splitlines()
N = int(input[0]) 
info = input[1:]
graph = list([] for _ in range(N + 1))
queue = deque([1])
parent = [0] * (N + 1) # parent[i]는 i의 부모, 0이면 미방문

# 인접리스트 만들기
for value in info:
    a, b = list(map(int, value.split()))
    graph[a].append(b)
    graph[b].append(a)

parent[1] = -1 # 루트 1은 부모 없음

# bfs
while queue:
    node = queue.popleft()

    for v in graph[node]:
        if parent[v] == 0: # 미방문
            parent[v] = node # v의 부모는 node
            queue.append(v)

result = []

for i in range(2, N + 1):
    result.append(str(parent[i]))

print("\n".join(result))   
